internal/users/tempentries/groups.go

```diff
@@ -0,0 +1,179 @@
+	registerMutex sync.Mutex
+	rwMutex       sync.RWMutex
+	groups        map[uint32]groupRecord
+	gidByName     map[string]uint32
+}
+
+func newTemporaryGroupRecords(idGenerator IDGenerator) *temporaryGroupRecords {
+	return &temporaryGroupRecords{
+		idGenerator:   idGenerator,
+		registerMutex: sync.Mutex{},
+		rwMutex:       sync.RWMutex{},
+		groups:        make(map[uint32]groupRecord),
+		gidByName:     make(map[string]uint32),
+	}
+}
+
+// GroupByID returns the group information for the given group ID.
+func (r *temporaryGroupRecords) GroupByID(gid uint32) (types.GroupEntry, error) {
+	r.rwMutex.RLock()
+	defer r.rwMutex.RUnlock()
+
+	group, ok := r.groups[gid]
+	if !ok {
+		return types.GroupEntry{}, NoDataFoundError{}
+	}
+
+	return r.groupEntry(group), nil
+}
+
+// GroupByName returns the group information for the given group name.
+func (r *temporaryGroupRecords) GroupByName(name string) (types.GroupEntry, error) {
+	r.rwMutex.RLock()
+	defer r.rwMutex.RUnlock()
+
+	gid, ok := r.gidByName[name]
+	if !ok {
+		return types.GroupEntry{}, NoDataFoundError{}
+	}
+
+	return r.GroupByID(gid)
+}
+
+func (r *temporaryGroupRecords) groupEntry(group groupRecord) types.GroupEntry {
+	return types.GroupEntry{Name: group.name, GID: group.gid, Passwd: group.passwd}
+}
+
+// RegisterGroup registers a temporary group with a unique GID in our NSS handler (in memory, not in the database).
+//
+// Returns the generated GID and a cleanup function that should be called to remove the temporary group once the group
+// was added to the database.
+func (r *temporaryGroupRecords) RegisterGroup(name string) (gid uint32, cleanup func() error, err error) {
+	r.registerMutex.Lock()
+	defer r.registerMutex.Unlock()
+
+	// Check if there is already a temporary group with this name
+	_, err = r.GroupByName(name)
+	if err != nil && !errors.Is(err, NoDataFoundError{}) {
+		return 0, nil, fmt.Errorf("could not check if temporary group %q already exists: %w", name, err)
+	}
+	if err == nil {
+		return 0, nil, fmt.Errorf("group %q already exists", name)
+	}
+
+	// Generate a GID until we find a unique one
+	for {
+		gid, err = r.idGenerator.GenerateGID()
+		if err != nil {
+			return 0, nil, err
+		}
+
+		// To avoid races where a group with this GID is created by some NSS source after we checked, we register this
+		// GID in our NSS handler and then check if another group with the same GID exists in the system. This way we
+		// can guarantee that the GID is unique, under the assumption that other NSS sources don't add groups with a GID
+		// that we already registered (if they do, there's nothing we can do about it).
+		var tmpID string
+		tmpID, cleanup, err = r.addTemporaryGroup(gid, name)
+		if err != nil {
+			return 0, nil, fmt.Errorf("could not register temporary group: %w", err)
+		}
+
+		unique, err := r.uniqueNameAndGID(name, gid, tmpID)
+		if err != nil {
+			if cleanupErr := cleanup(); cleanupErr != nil {
+				err = errors.Join(err, cleanupErr)
+			}
```

Comment:
Could it just be `errors.Join(err, cleanup())`?

Reply:
It could, but I would find that harder to read. When you read that, you need to know and remember that `errors.Join` ignores `nil` arguments.